### ISSUE TITLE
issue with url-encoded path segments on server-side

### ISSUE THREAD

**Issue by h0ru5** [eclipse/californium#5]
If you are unlucky enough to find yourself in the situation that you get non-ascii chars or even blanks in the path, its crucial to uri-decode the path segments. Quick fix:

replace line:
https://github.com/eclipse/californium/blob/master/californium-core/src/main/java/org/eclipse/californium/core/server/ServerMessageDeliverer.java#L142

by something like:

``` java
String name = path.removeFirst();
String sanitizedName = java.net.URLDecoder.decode(name, "UTF-8");
current = current.getChild(sanitizedName);
```

**Comment by mkovatsc** [comment]
I remember this issue at a different point, where it was actually the URI class to cause the problem. Not sure if this is the right place, since it is simply a .get() on a ConcurrentHashMap, which does not know about URI encoding.

I will look into this and try to fix it at the root...

For 2.x, we will need to have something better than Java's own URI class anyway to support alternative transports.

**Comment by h0ru5** [comment]
problem is here for example: you have a uri path token like ``stränge name``, which will be plaintext in the hashmap, but uri-encoded to ``str%C3%A4nge+name``  or  ``str%c3%a4nge+name`` in the request (or even ``str%C3%A4nge%20name`` when e.g. sent from copper). 

Therefore they will not match. I tried a workaround by injecting encoded keys, but there are just too many combinations. (... so I resorted to using english names :grin:)

**Comment by mkovatsc** [comment]
Yes, the problem is Java's URI class in Californium. We should do clean UTF-8 handling.

Copper has to clean what it gets from Firefox. I will have a look there in parallel when fixing Cf.

Luckily it is hard to inject stränge stringß in Erbium ;)
(which would be handled perfectly as UTF-8 bytes---I hope)

**Comment by h0ru5** [comment]
Actually in the pointed code its just plain Strings, not the Java URI class.
The lines should fix the problem - just had not launched a PR yet

**Comment by mkovatsc** [comment]
What I mean is that at this point the string should already be exactly what it is supposed to mean. It must be identical to the UTF-8-encoded bytes in the Uri-Option. If the Uri-Option already carried an encoded string, the client has to be fixed. For Cf, we have to make sure, we do not convert any URI option into an encoded string. The later, however, might happen when using the URI class to handle the full URI (e.g., sanity checks, normalizing, etc.).

I know of a similar problem when creating resources with a whitespace in the name.

**Comment by sophokles73** [comment]
@mkovatsc,

you  mentioned that

>For 2.x, we will need to have something better than Java's own URI class anyway to support alternative transports.

and

>Yes, the problem is Java's URI class in Californium. We should do clean UTF-8 handling.

I cannot really see what the problem with `java.net.URI` should be in this context. We may (currently) not use this class in an appropriate way but my understanding of `URI`'s JavaDoc suggests that it is perfectly suited for being used for encoding/decoding URIs within Californium.

My gut feeling is that our current usage of plain strings for storing URIs within e.g. `CoapClient` is the source of any problems in this area whereas consequent usage of the `URI` class for that matter would actually provide us with a consistent approach to encoding/escaping characters ...

**Comment by mkovatsc** [comment]
CoAP does support clean UTF-8 URIs and does not require, and hence should not use any escaping. When we start to have elements in between that introduce any kind of encoding, we end up with exactly the mess we have at the moment. It would, for instance, also get very messy with upper vs lower case of percent encodings. I also had some other trouble with the URI class it some point; unfortunately, I cannot put my finger on it at the moment.

**Comment by sophokles73** [comment]
I have added some test cases regarding this issue to `RequestTest`. However, I couldn't find a problem with Californium's handling of options and URI strings so I am now closing this issue.
Please create a new issue with a concrete example of unexpected behavior if you still experience problems.
